compiler/rustc_mir_transform/src/scalar_deref_prop.rs

```diff
@@ -0,0 +1,426 @@
+use either::Either;
+use rustc_index::bit_set::BitSet;
+use rustc_index::vec::IndexVec;
+use rustc_middle::middle::resolve_lifetime::Set1;
+use rustc_middle::mir::visit::*;
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+
+use crate::MirPass;
+
+/// Propagate references inside a basic block.
+///
+/// MIR building may produce a lot of borrow-dereference patterns.
+///
+/// This pass aims to transform the following pattern:
+///   _1 = &raw? mut? _2;
+///   _3 = *_1;
+///
+/// Into
+///   _1 = _2;
+///
+/// For `&mut` borrows, we need to preserve the uniqueness property:
+/// we must avoid creating a state where we interleave uses of `*_1` and `_2`.
+/// To do it, we only perform full substitution of mutable borrows:
+/// we replace either all or none of the occurrences of `*_1`.
+///
+/// Some care has to be taken when `_1` is copied in other locals.
+///   _1 = &raw? mut? _2;
+///   _3 = *_1;
+///   _4 = _1
+///   _5 = *_4
+/// In such cases, fully substituting `_1` means fully subtituting all of the copies.
+///
+/// For immutable borrows, we do not need to preserve such uniqueness property,
+/// so we perform all the possible substitutions without removing the `_1 = &_2` statement.
+///
+/// For simplicity, we treat raw pointers exactly like references.  This is sound because
+/// references impose strictly more constraints than raw pointers.
+pub struct PropagateReferences;
+
+impl<'tcx> MirPass<'tcx> for PropagateReferences {
+    fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
+        sess.mir_opt_level() >= 4
+    }
+
+    #[instrument(level = "trace", skip(self, tcx, body))]
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        debug!(def_id = ?body.source.def_id());
+        propagate_ssa(tcx, body);
+    }
+}
+
+fn propagate_ssa<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    let mut ssa = SsaLocals::new(body);
+
+    let copy_classes = ssa.compute_copy_classes(body);
+    debug!(?copy_classes);
+
+    let mut fully_replacable_locals = fully_replacable_locals(&ssa, &copy_classes);
+    debug!(?fully_replacable_locals);
+
+    let targets = compute_local_values(tcx, body, &ssa, &fully_replacable_locals);
+    debug!(?targets);
+
+    let mut storage_to_remove = BitSet::new_empty(body.local_decls.len());
+    for (local, v) in targets.iter_enumerated() {
+        match v {
+            Value::Pointer(p) => {
+                if p.projection.first() == Some(&PlaceElem::Deref) {
+                    // We may introduce uses of a local after its `StorageDead`.
+                    // To avoid introducing UB, just remove the StorageDead statements.
```

Comment:
We may have a silly ICE here. If the original code is UB, like
```rust
fn ub() {
    let x = {
        let y = vec![8_usize];
        std::ptr::addr_of!(y)
    };
    let z = unsafe { &*x };
    opaque(z);
}
```

Then the optimization will replace the reborrow by `let z = &y`. This is the same UB, but this time MIR validation will ICE because `y` is used outside of storage liveness range.

I see 4 options:
- remove the check from MIR validation (bad idea, it's a very useful check);
- remove the `StorageDead` in more cases (= remove some UB);
- replace the UB by a `Assert` terminator explaining the UB;
- replace the UB by a `Unreachable` terminator.

What do you think?

Reply:
Ideally we'd make it an assert with debug assertions and an unreachable in release mode. Similar to how there's a PR for debug assertions on raw pointer derefs, checking for correct alignment